### DISCUSSION TITLE
Powershell check (within echoc)

### DISCUSSION THREAD

**Discussion by DavidGeeraerts** [DarviL82/DarviLStuff#1] [General]
Curious about your thoughts on checking for powershell:
I see that echoc is using registry:

```
::Check if the user has PowerShell installed.
	<nul set /p =Checking PowerShell... 
	reg query HKLM\SOFTWARE\Microsoft\PowerShell\1 /v Install 1> "%temp%/.tmp" 2>&1 & find "Install    REG_DWORD    0x1" "%temp%/.tmp" > nul
	if !errorlevel!==1 (
		call :display red "PowerShell isn't installed. Altough, it isn't required."
	) else call :display green "PowerShell is installed."
```

Here's are the ways to check for powershell (that I know of):

1. `@powershell (Get-Host).version`
2. `@powershell $PSVersionTable.PSVersion`
2. `if defined PSModulePath ...`
3. Registry (as coded above)

- Preference for speed, e.g.  `if defined PSModulePath ...` might be the fastest (?), but wouldn't tell you that powershell is actually working.
- Let's say a program is dependent on powrshell, e.g. hybrid program using both cmd & powershell (mostly called via @powershell), then actually testing powershell with either of these 2 commands would be better: `@powershell (Get-Host).version`  , `@powershell $PSVersionTable.PSVersion`

(On a side note, I often use `@powershell Write-Host "Searching..." -ForegroundColor DarkYellow` to get a line of string as a different color.)

_I've submitted this mostly as a feeler if you might be interested in chatting about windows command shell._

**Comment by darvil82** [comment]
Hello! Thank you for taking your time helping me, nice to talk with you!
First of all, I will change the way of checking for Powershell with your first suggestion, feels definitely better in all ways, since I don't really care about speed in this case. By the way, is there any difference between `@powershell (Get-Host).version` and `@powershell $PSVersionTable.PSVersion`? Seems like they work the same for me.
Also, this script used to only use Powershell when displaying lines with colors, but then I added the ability of using the ANSI escape codes, because it is faster than calling powershell (which was a problem when displaying lines, due to the slow speed and lots of errors with special characters), and accepts various colors in one line.

Greetings.
Pd: I didn't expect an issue in here hehe.

**Comment by darvil82** [comment]
I'm assuming that a simple `ERRORLEVEL` check would be enough right?
```bat
	::Check if the user has PowerShell installed.
	<nul set /p =Checking PowerShell... 
	powershell ^(Get-Host^).version > nul
	if not !errorlevel!==0 (
		call :display red "PowerShell isn't installed. Altough, it isn't required."
	) else call :display green "PowerShell is installed."
```

**Comment by DavidGeeraerts** [comment]
Regarding `Get-Host` _vs_ `$PSVersionTable`:

`Get-host` is higher level info stuff.
` $PSVersionTable` is a powershell variable that has specific powershell version information.

I prefer ` $PSVersionTable` because it is specific.
Here's what I do to parse the info:

```
echo Checking on Powershell...
IF EXIST "%LOG_LOCATION%\PS_Version.txt" goto skipChkPS
IF NOT DEFINED PSModulePath GoTo skipChkPS
@powershell $PSVersionTable.PSVersion > "%LOG_LOCATION%\PS_Version.txt"
:checkPS
FOR /F "usebackq skip=3 tokens=1 delims= " %%P IN ("%LOG_LOCATION%\PS_Version.txt") DO SET "$PS_MAJOR_VERSION=%%P"
FOR /F "usebackq skip=3 tokens=2 delims= " %%P IN ("%LOG_LOCATION%\PS_Version.txt") DO SET "$PS_MINOR_VERSION=%%P"
FOR /F "usebackq skip=3 tokens=3 delims= " %%P IN ("%LOG_LOCATION%\PS_Version.txt") DO SET "$PS_BUILD_VERSION=%%P"
FOR /F "usebackq skip=3 tokens=4 delims= " %%P IN ("%LOG_LOCATION%\PS_Version.txt") DO SET "$PS_REVISION_VERSION=%%P"
:skipChkPS

```

Yes, using powershell to control string color in console is slower then using native console. This is one of the reasons I'm looking at how echoc is controlling color.

Capturing `ERRORLEVEL` is what I would also do with that section of code you wrote.

**Comment by darvil82** [comment]
Alright, I will use the variable method then. By the way, I noticed that you are calling `FOR` four times to parse everything, but you could do it with just one call like this:
```bat
:checkPS
FOR /F "usebackq skip=3 tokens=1-4 delims= " %%P IN ("%LOG_LOCATION%\PS_Version.txt") DO SET "$PS_MAJOR_VERSION=%%P" & set "$PS_MINOR_VERSION=%%Q" & set "$PS_BUILD_VERSION=%%R" & set "$PS_REVISION_VERSION=%%S"
:skipChkPS
```
`FOR` has the ability of having one token per variable, so by incrementing %%P to the next letter you could define all of those variables in one shot.

**Comment by DavidGeeraerts** [comment]
Thanks for the catch/tip on the FOR.
I know I've done that with some other code --think I get lost when writing lots of code.

Here's me practicing pretty code: 

```
:checkPS
FOR /F "usebackq skip=3 tokens=1-4 delims= " %%P IN ("%LOG_LOCATION%\PS_Version.txt") DO (
	SET "$PS_MAJOR_VERSION=%%P") & (
	SET "$PS_MINOR_VERSION=%%Q") & (
	SET "$PS_BUILD_VERSION=%%R") & (
	SET "$PS_REVISION_VERSION=%%S")
	)
:skipChkPS
```

**Comment by darvil82** [comment]
Neat! In that case you could simplify it by just doing this:
```bat
:checkPS
FOR /F "usebackq skip=3 tokens=1-4 delims= " %%P IN ("%LOG_LOCATION%\PS_Version.txt") DO (
	SET "$PS_MAJOR_VERSION=%%P"
	SET "$PS_MINOR_VERSION=%%Q"
	SET "$PS_BUILD_VERSION=%%R"
	SET "$PS_REVISION_VERSION=%%S"
)
:skipChkPS
```
It won't change anything really, but it will help a bit for making it cleaner 👍

**Comment by DavidGeeraerts** [comment]
Oh, very cool!
Like I said, I need practice doing pretty code, and that makes it much prettier!

**Comment by darvil82** [comment]
Alright! Take your time then. I will be closing this issue now, and I will convert it into a discussion. (just so I can also test this new interesting feature haha)
Let me know if you have any question! 😃
